README.md
# bam2bedpe

## Summary
- scripts for converting BAM into BEDPE format
- BAM file size prohibitive large
- BEDPE retains essential BAM information while having 6x size reduction
- bedtools bam2bedpe conversion cannot handle secondary alignments
- bam2bedpe script here preserves all alignments

## Workflow overview

![bam2bedpe](https://user-images.githubusercontent.com/98410560/170511327-ca6435bf-9d32-4676-890e-caf9f067f1ad.png)

## BEDPE7+12 file format
- inspired by bedtools definition (https://bedtools.readthedocs.io/en/latest/content/general-usage.html)
- originally designed to describe disjoint genome features (e.g. structural variants), also used for paired end reads which cannot be handled by traditional BED files
- first read in pair is always READ1, i.e. column order is READ1 READ2 (as oppose to coordinate order)
- follows 0-based coordinate system ( first base in chromosome is numbered 0 )
- chromosome name can be any character ( e.g. chr6_KQ090016v1_fix, chr10_GL383545v1_alt, chr11_KI270721v1_random, chrUn_GL000195v1, scaffold123, contig111 )
- TLEN is identical to ones found in BAM (calculated where length only include M CIGAR)

### BEDPE7+12 fields

| Col  | BED Field        | Type   | Regex or range           | Brief Description                | BAM Equivalent            |
| ---- | ---------------- | ------ | ------------------------ | -------------------------------- | ------------------------- |
| 1    | read1_chrom      | String | `.\[[:alnum:]_]{1,255}^4`| read1 chromosome name            | RNAME                     |
| 2    | read1_chromStart | Int    | `-1\|([0,2^64−1])`       | read1 start position             | POS                       |
| 3    | read1_chromEnd   | Int    | `-1\|([0,2^64−1])`       | read1 end position               | None, calculated by pysam |
| 4    | read2_chrom      | String | `[[:alnum:]_]{1,255}^4`  | read2 chromosome name            | RNAME                     |
| 5    | read2_chromStart | Int    | `-1\|([0,2^64−1])`       | read2 start position             | POS                       |
| 6    | read2_chromEnd   | Int    | `-1\|([0,2^64−1])`       | read2 end position               | None, calculated by pysam |
| 7    | fragment_id      | String | `[\x20-\x7e]{1,255}`     | fragment id                      | QNAME                     |
| 8    | read1_MAPQ       | Int    | `[0, 2^8-1]`             | read1 mapping quality            | MAPQ                      |
| 9    | read2_MAPQ       | Int    | `[0, 2^8-1]`             | read2 mapping quality            | MAPQ                      |
| 10   | read1_STRAND     | String | `[-+.]`                  | read1 strand                     | None, calculated by pysam |
| 11   | read2_STRAND     | String | `[-+.]`                  | read2 strand                     | None, calculated by pysam |
| 12   | read1_CIGAR      | String | `*\|([0-9]+[MIDNSHPX=])+`| read1 CIGAR string               | CIGAR                     |
| 13   | read2_CIGAR      | String | `*\|([0-9]+[MIDNSHPX=])+`| read2 CIGAR string               | CIGAR                     |
| 14   | read1_FLAG       | Int    | `[0, 2^16 -1]`           | read1 bitwise FLAG               | FLAG                      |
| 15   | read2_FLAG       | Int    | `[0, 2^16 -1]`           | read2 bitwise FLAG               | FLAG                      |
| 16   | read1_TLEN       | Int    | `[-2^31 + 1, 2^31 -1]`   | read1 template length            | TLEN                      |
| 17   | read2_TLEN       | Int    | `[-2^31 + 1, 2^31 -1]`   | read2 template length            | TLEN                      |
| 18   | read1_NM_TAG     | Int    | `NA\|([0-9]*)`           | read1 edit distance to reference | NM                        |
| 19   | read2_NM_TAG     | Int    | `NA\|([0-9]*)`           | read2 edit distance to reference | NM                        |

Please see [wiki](https://github.com/mhanbioinfo/bam2bedpe/wiki) for tutorial on settings up and running module.

Questions please contact ming.han@uhn.ca.


CHANGELOG
2022-06-06: bam2bedpe module (version 0.2.0)
- working bedpe7+12 file validator, example valid/invalid bedpe.gz files

2022-05-25: bam2bedpe module (version 0.1.0)
- working bam2bedpe module, 
- conda env installation instructions


run_bam2bedpe.sh
#!/bin/bash

#SBATCH --mail-type=ALL
#SBATCH --mail-user=ming.han@uhn.ca
#SBATCH -D ./logs_slurm/
#SBATCH -t 1-00:00:00
#SBATCH --mem=16G
#SBATCH -J bam2bedpe
#SBATCH -p himem
#SBATCH -c 1
#SBATCH -N 1
#SBATCH -o %j-%x.out
#SBATCH -e %j-%x.err

if [ $# != 1 ]; then
    echo "Please specify a config.yml file."
    exit 1
fi

## PARSE CONFIG ######################################################

## https://github.com/ash-shell/yaml-parse/blob/master/lib/yaml_parse.sh
## yml file must be 2 spaces indented
function parse_yaml {
   local prefix=$2
   local s='[[:space:]]*' w='[a-zA-Z0-9_]*' fs=$(echo @|tr @ '\034')
   sed -ne "s|^\($s\):|\1|" \
        -e "s|^\($s\)\($w\)$s:$s[\"']\(.*\)[\"']$s\$|\1$fs\2$fs\3|p" \
        -e "s|^\($s\)\($w\)$s:$s\(.*\)$s\$|\1$fs\2$fs\3|p"  $1 |
   awk -F$fs '{
      indent = length($1)/2;
      vname[indent] = $2;
      for (i in vname) {if (i > indent) {delete vname[i]}}
      if (length($3) > 0) {
         vn=""; for (i=0; i<indent; i++) {vn=(vn)(vname[i])("_")}
         gsub(/\s*#.*$/, "", $3);
         printf("%s%s%s=\"%s\"\n", "'$prefix'",vn, $2, $3);
      }
   }'
}

CONFIG_F=$1
params=($(parse_yaml $CONFIG_F))

for param in "${params[@]}"; do
   #echo $param
   lhs=${param%=*}
   rhs=${param#*=}
   rhs="${rhs%%\#*}"    # Del in line right comments
   rhs="${rhs%%*( )}"   # Del trailing spaces
   rhs="${rhs%\"*}"     # Del opening string quotes 
   rhs="${rhs#\"*}"     # Del closing string quotes 
   declare $lhs=$rhs
done


## Main program ####################################################

source ${CONDA_ACTIVATE} ${CONDA_ENV}

echo "Processing bam2bedpe..."
echo "output path:      $OUT_DIR"
echo "number of chunks: $NUM_OF_CHUNKS"
echo "processing on:    $SLURMLOCAL"
echo ""

SRC_DIR="$(pwd)"

OLDIFS=$IFS
IFS=','
[ ! -f $SAMPLESHEET_PATH ] && { echo "$SAMPLESHEET_PATH file not found"; exit 99; }
sed 1d $SAMPLESHEET_PATH | \
while read -r SAMPLE_NAME INPUT_BAM_PATH; do


    #SAMPLE_NAME="toy01"
    #INPUT_BAM_PATH="/cluster/home/t110409uhn/git/cfmedip_medremix_bedpe_git/toy_files/toy01.bam"
    
    echo "Processing sample: ${SAMPLE_NAME}"
    echo "Input bam path:    ${INPUT_BAM_PATH}"
    echo "Job started at "$(date) 
    time1=$(date +%s)
    
    PY_SCRIPT_DIR=${SRC_DIR}
    PY_SCRIPT_PATH="${PY_SCRIPT_DIR}/bam2bedpe_pysam.py"
    INPUT_DIR="${INPUT_BAM_PATH%/*}"
    INPUT_BAM="${INPUT_BAM_PATH##*/}"
    
    TMP_DIR="${OUT_DIR}/tmp_dir/${INPUT_BAM%.*}"
    mkdir -p $TMP_DIR
    
    OUT_FRAG_NAMES="${INPUT_BAM%.*}.fragNames"
    OUT_MERGED_SORTD_BEDPE="${INPUT_BAM%.*}_coordSortd.bedpe"
    
    ## -------------------------------------- ##
    ## get fragNames
    samtools view ${INPUT_DIR}/${INPUT_BAM} \
        | cut -f1 \
        | awk '{ a[$1]++ } END { for (b in a) { print b } }' \
        > ${TMP_DIR}/${OUT_FRAG_NAMES}
    
    ## -------------------------------------- ##
    ## split bam into chunks
    NUM_ROWS=$(cat ${TMP_DIR}/${OUT_FRAG_NAMES} | wc -l)
    CHUNK_SIZE=$(( $NUM_ROWS / $NUM_OF_CHUNKS ))   ## won't have decimal, sometimes will be short, last chunk has to go to last row
    echo "bam file has $NUM_ROWS fragments."
    echo "Number of chunks was set to ${NUM_OF_CHUNKS}."
    echo "Each chunk will be $CHUNK_SIZE rows."
    
    for ((i=0; i<=$(( $NUM_OF_CHUNKS - 2 )); i++)); do
        CHUNK=$(( i + 1 ))
        ROW_START=$(( i*CHUNK_SIZE + 1))
        ROW_END=$(( CHUNK*CHUNK_SIZE ))
        echo "Processing CHUNK${CHUNK}... starting with row ${ROW_START}, ending in row ${ROW_END}."
        sed -n "$ROW_START,$ROW_END p" ${TMP_DIR}/${OUT_FRAG_NAMES} \
            > ${TMP_DIR}/${OUT_FRAG_NAMES}.CHUNK${CHUNK}
    
    done
    
    ## since chunk might not be evenly divided, last chunk will just be to last row
    ith=$(( $NUM_OF_CHUNKS - 1 ))
    ROW_START=$(( ith*CHUNK_SIZE + 1 ))
    echo "Processing CHUNK$NUM_OF_CHUNKS... starting with row ${ROW_START}, ending in row ${NUM_ROWS}."
    sed -n "$ROW_START,$NUM_ROWS p" ${TMP_DIR}/${OUT_FRAG_NAMES} \
        > ${TMP_DIR}/${OUT_FRAG_NAMES}.CHUNK${NUM_OF_CHUNKS}
    
    ## make log dir
    mkdir -p ${OUT_DIR}/logs_slurm
    
    ## -------------------------------------- ##
    ## picard FilterSamReads on chunks
    for CHUNK in ${TMP_DIR}/${OUT_FRAG_NAMES}.CHUNK*; do
        echo "Picard FilterSamReads for ${CHUNK##*/}..."
        echo "Output is ${INPUT_BAM%.*}_${CHUNK##*.}.bam"
        
        echo "Writing out ${INPUT_BAM%.*}_bam2bedpe_${CHUNK##*.}.sh"
        ## write out sample sbatch script
        cat <<- EOF > "${TMP_DIR}/${INPUT_BAM%.*}_bam2bedpe_${CHUNK##*.}.sh" 
	#!/bin/bash
	#SBATCH -t 3-00:00:00
	#SBATCH -J bam2bedpe_chunks_${CHUNK##*.}
	#SBATCH -D ${OUT_DIR}/logs_slurm
	#SBATCH --mail-type=ALL
	#SBATCH --mail-user=ming.han@uhn.ca
	#SBATCH -p himem
	#SBATCH -c 1
	#SBATCH --mem=16G
	#SBATCH -o ./%j-%x.out
	#SBATCH -e ./%j-%x.err
	
	source ${CONDA_ACTIVATE} ${CONDA_ENV}
	
	echo "Job started at "\$(date) 
	time1=\$(date +%s)
	
	picard FilterSamReads \
	    -I ${INPUT_DIR}/${INPUT_BAM} \
	    -O ${TMP_DIR}/"${INPUT_BAM%.*}_${CHUNK##*.}.bam" \
	    -READ_LIST_FILE ${CHUNK} \
	    -FILTER includeReadList \
	    -WRITE_READS_FILES false \
	    -USE_JDK_DEFLATER true \
	    -USE_JDK_INFLATER true \
	
	python ${PY_SCRIPT_PATH} \
	    --sort_bam_by_qname \
	    --bam_input ${TMP_DIR}/"${INPUT_BAM%.*}_${CHUNK##*.}.bam" \
	    --bedpe_output ${TMP_DIR}/"${INPUT_BAM%.*}_${CHUNK##*.}.bedpe"
	
	rm ${TMP_DIR}/"${INPUT_BAM%.*}_${CHUNK##*.}.bam"
	
	time2=\$(date +%s)
	echo "Job ended at "\$(date) 
	echo "Job took \$(((time2-time1)/3600)) hours \$((((time2-time1)%3600)/60)) minutes \$(((time2-time1)%60)) seconds"
	EOF
    
        if [ $SLURMLOCAL == "slurm" ]; then
            sbatch "${TMP_DIR}/${INPUT_BAM%.*}_bam2bedpe_${CHUNK##*.}.sh"
        elif [ $SLURMLOCAL == "local" ]; then
            nohup bash "${TMP_DIR}/${INPUT_BAM%.*}_bam2bedpe_${CHUNK##*.}.sh" &> "${TMP_DIR}/${INPUT_BAM%.*}_bam2bedpe_${CHUNK##*.}.log" &
        fi
    
    done
    
    echo $TMP_DIR
    echo $OUT_DIR
    echo $OUT_MERGED_SORTD_BEDPE
    
    
    ## periodically check if chunks have been written to completely
    while true; do
        if [ $(find ${TMP_DIR} -maxdepth 1 -mmin +3 -type f -regex ".*_CHUNK[1-9][0-9]*\.bedpe" | wc -l) -eq $NUM_OF_CHUNKS ] 
        then
            echo "All bedpe chunks have been written to, merging bedpe chunks..."
            find ${TMP_DIR} -maxdepth 1 -mmin +3 -type f -regex ".*_CHUNK[1-9][0-9]*\.bedpe" -exec cat {} + \
                | sort -k1,1V -k2,2n -k3,3n -k5,5n -k6,6n \
                | gzip -c \
                > ${OUT_DIR}/${OUT_MERGED_SORTD_BEDPE}.gz
            break 1
        else
            echo "Sleeping for 5 minutes, then recheck if bedpe chunks were written to completely."
            sleep 5m
        fi
    done
    
    ## remove TMP_DIR
    if "$KEEP_TMP"; then
        echo "Keeping temp dir"
    else
        echo "Removed temp dir after completion"
        rm -rf ${TMP_DIR}
    fi
    
    
    time2=$(date +%s)
    echo "Job ended at "$(date) 
    echo "Job took $(((time2-time1)/3600)) hours $((((time2-time1)%3600)/60)) minutes $(((time2-time1)%60)) seconds"
    echo ""


done
IFS=$OLDIFS


echo "Finished processing bam2bedpe for all samples."


bam2bedpe_pysam_v5_wCIGAR.py
#!/usr/bin/env python
## bam2bedpe with FLAG, NM, TLEN
## USAGE:
## python bam2bedpe_pysam_v?.py [--sort_bam_by_qname] --bam_input --bedpe_output

import os
import pysam
import argparse
from copy import deepcopy
from timeit import default_timer as timer
import time

parser = argparse.ArgumentParser()
parser.add_argument('--sort_bam_by_qname', help='Will start with `samtools sort -n ...`', action='store_true')
parser.add_argument('--bam_input', help='Specify full path to BAM file.', required=True)
parser.add_argument('--bedpe_output', help='Specify full path of bedpe file.', required=True)
args = parser.parse_args()

sort_bam = args.sort_bam_by_qname
bam_input_path = args.bam_input
bedpe_output_path = args.bedpe_output

output_dir = os.path.dirname(bedpe_output_path)

## sort bam by queryname
if sort_bam is True:
    print("Sorting bam by queryname... ")
    bam_namedSortd_path = os.path.join(output_dir, os.path.splitext(os.path.basename(bam_input_path))[0] + "_nameSortd.bam")
    # print(bam_namedSortd_path)
    pysam.sort(bam_input_path, "-n", "-o", bam_namedSortd_path)
    bam_input_path = bam_namedSortd_path


## create new bedpe file
open(bedpe_output_path, 'w').close()

## start timer
start = timer()

## ------------------------------------------------------------ ##
## function to write out bedpe from bam
def write_bedpe(frag_dict):
    for key, an_fragmnt in frag_dict.items():

        if len(an_fragmnt) == 1:
            print("Only 1 read for fragment: ", key)

        bedpe_file_out = open(bedpe_output_path, "a")
        # a_fragmnt = frags_dict['NB551051:196:H3GGWBGXH:1:12104:6290:3128']
        
        ## deal with flags 329_393_377_441 reads
        a_fragmnt_alignmts_lst = [] ## for removing duplicate lines later
        for read_a in an_fragmnt:
            
            ## deal with flags 329_393_377_441 reads
            if read_a.flag in [329,393,377,441]:
                a_fragmnt_alignmts_lst.append("".join(str(x) for x in [ \
                    read_a.reference_name, "\t", read_a.reference_start, "\t", read_a.reference_start, "\t", \
                    read_a.reference_name, "\t", read_a.reference_start, "\t", read_a.reference_start, "\t", \
                    read_a.query_name, "\t", read_a.mapping_quality, "\t", read_a.mapping_quality, "\t", \
                    ".", "\t", ".", read_a.cigarstring, "\t", read_a.cigarstring, "\t", \
                    read_a.flag, "\t", read_a.flag, "\t", \
                    read_a.template_length, "\t", read_a.template_length, "\t",
                    "NA", "\t", "NA", "\n"]))

            for read_b in an_fragmnt:
                if read_a != read_b:
                    if read_a.next_reference_name == read_b.reference_name and \
                    read_a.next_reference_start == read_b.reference_start:
                        
                        ## define chr, chr_end if unmapped
                        if read_a.is_unmapped:
                            read_a_ref_name = "."
                            read_a_ref_start = "-1"
                            read_a_ref_end = "-1"
                            read_a_cigar = "*"
                        else:
                            read_a_ref_name = read_a.reference_name
                            read_a_ref_start = read_a.reference_start
                            read_a_ref_end = read_a.reference_end
                            read_a_cigar = read_a.cigarstring
                        if read_b.is_unmapped:
                            read_b_ref_name = "."
                            read_b_ref_start = "-1"
                            read_b_ref_end = "-1"
                            read_b_cigar = "*"
                        else:
                            read_b_ref_name = read_b.reference_name
                            read_b_ref_start = read_b.reference_start
                            read_b_ref_end = read_b.reference_end
                            read_b_cigar = read_b.cigarstring
                        
                        ## define strand
                        if read_a.is_reverse: read_a_strand = "-"
                        else: read_a_strand = "+"
                        if read_a_ref_name == ".": read_a_strand = "."

                        if read_b.is_reverse: read_b_strand = "-"
                        else: read_b_strand = "+"
                        if read_b_ref_name == ".": read_b_strand = "."
                        
                        ## define NM tag
                        if 'NM' in dict(read_a.tags): read_a_NM_tag = dict(read_a.tags)['NM']
                        else: read_a_NM_tag = "NA"
                        if 'NM' in dict(read_b.tags): read_b_NM_tag = dict(read_b.tags)['NM']
                        else: read_b_NM_tag = "NA"
                        
                        ## read1 always in col1-3, same as bedtools bedpe -mate1
                        if read_a.is_read1:
                            # pass
                            a_fragmnt_alignmts_lst.append("".join(str(x) for x in [ \
                                read_a_ref_name, "\t", read_a_ref_start, "\t", read_a_ref_end, "\t", \
                                read_b_ref_name, "\t", read_b_ref_start, "\t", read_b_ref_end, "\t", \
                                read_a.query_name, "\t", read_a.mapping_quality, "\t", read_b.mapping_quality, "\t", \
                                read_a_strand, "\t", read_b_strand, "\t", \
                                read_a_cigar, "\t", read_b_cigar, "\t", \
                                read_a.flag, "\t", read_b.flag, "\t", \
                                read_a.template_length, "\t", read_b.template_length, "\t",
                                read_a_NM_tag, "\t", read_b_NM_tag, "\n"]))
                        elif read_b.is_read1:
                            # pass
                            a_fragmnt_alignmts_lst.append("".join(str(x) for x in [ \
                                read_b_ref_name, "\t", read_b_ref_start, "\t", read_b_ref_end, "\t", \
                                read_a_ref_name, "\t", read_a_ref_start, "\t", read_a_ref_end, "\t", \
                                read_b.query_name, "\t", read_b.mapping_quality, "\t", read_a.mapping_quality, "\t", \
                                read_b_strand, "\t", read_a_strand, "\t", \
                                read_b_cigar, "\t", read_a_cigar, "\t", \
                                read_b.flag, "\t", read_a.flag, "\t", \
                                read_b.template_length, "\t", read_a.template_length, "\t",
                                read_b_NM_tag, "\t", read_a_NM_tag, "\n"]))
                        
        for line in list(set(a_fragmnt_alignmts_lst)):
            # print(line)
            bedpe_file_out.write(line)
        bedpe_file_out.close()

## ------------------------------------------------------------ ##
## iterate over pysam generator

## Initial call to print 0% progress
#samfile1 = pysam.AlignmentFile(bam_input_path) ## just to get len of generator
#loop_len = len(list(samfile1))
#printProgressBar(0, loop_len, prefix = 'Progress:', suffix = 'Complete', length = 50)

## read in pysam generator
samfile = pysam.AlignmentFile(bam_input_path)

qname_previous = ""
frag_dict = {}
frag_next_dict = {}

## loop through generator
for i, line in enumerate(samfile):
    
    qname = line.query_name
    # print(qname)
    
    ## just to get it started
    if qname_previous == "":
        frag_dict[qname] = [line]
        qname_previous = deepcopy(qname)
        continue
    
    ## compare current line to previous line
    if qname == qname_previous:
        frag_dict[qname].append(line)
        qname_previous = deepcopy(qname)
    
    ## as soon as current line qname != previous line qname,
    ## save iter to next fragment dict, and
    ## execute write_bedpe()
    else:
        frag_next_dict[qname] = [line]
        # print(frag_next_dict)
        
        write_bedpe(frag_dict)
        
        qname_previous = deepcopy(qname)
        frag_dict = deepcopy(frag_next_dict)
        frag_next_dict = {}
    
## write final fragment
write_bedpe(frag_dict)

## ------------------------------------------------------------ ##
## sort bedpe by coordinates
#bedpe_sortd_path = os.path.splitext(bedpe_output_path)[0] + "_coordSortd.bedpe"
#os.system("sort -k1,1V -k2,2n -k3,3n -k5,5n -k6,6n %s > %s" % (bedpe_output_path, bedpe_sortd_path))

## ------------------------------------------------------------ ##

end = timer()
#print("Processing bam2bedpe took ", str(end - start), " seconds.")


## EOF


bedpe7plus12validation.sh
#!/bin/bash

#SBATCH --mail-type=ALL
#SBATCH --mail-user=ming.han@uhn.ca
#SBATCH -t 1-00:00:00
#SBATCH -D ./logs_slurm/
#SBATCH --mem=16G
#SBATCH -J bedpe7plus12validation
#SBATCH -p himem
#SBATCH -c 1
#SBATCH -N 1
#SBATCH -o %j-%x.out
#SBATCH -e %j-%x.err

# getopts ###################################################
usage(){
    echo 
    echo "Usage: bash bedpe7plus12validation.sh -i BEDPE_INPUT"
    echo 
}
no_args="true"

## Help 
Help()
{
    # Display Help
    echo 
    echo "Checks if .bedpe.gz follows BEDPE7+12 format (including correct number of columns,tab-delimited,appropriate values for each column)."
    echo
    echo "Usage: bash bedpe7plus12validation.sh -i BEDPE_INPUT"
    echo "options:"
    echo "-h   [HELP]      print help"
    echo "-i   [REQUIRED]  input .bedpe.gz file (full path)"
    echo
}

## Get the options
while getopts ":hi:" option; do
    case "${option}" in
        h) Help
           exit;;
        i) INPUT_BEDPE=${OPTARG};;
       \?) echo "Error: Invalid option"
           exit;;
    esac
    no_args="false"
done

[[ "$no_args" == "true" ]] && { usage; exit 1; }

echo "Input .bedpe.gz: $INPUT_BEDPE"

# Main program ##############################################

echo "Job started at "$(date) 
time1=$(date +%s)

NUM_COLS=19
NUM_LINES=$(zcat ${INPUT_BEDPE} | wc -l)
#echo $NUM_LINES

LINE_NUM_NF_neq19=$(zcat ${INPUT_BEDPE} | awk -F'\t' 'BEGIN { ORS="," } NF != 19 {print NR}')
if [ ${#LINE_NUM_NF_neq19} -eq 0 ]; then
    echo "All rows have required 19 fields, tab-delimited."
else
    echo "These rows do not have the correct number of fields, and/or is not tab-delimited: ${LINE_NUM_NF_neq19%,*}."
fi

BAD_CHR_ROWS=$(zcat ${INPUT_BEDPE} | awk -F'\t' 'BEGIN { ORS="," } !(($1 ~/[A-Za-z]+[1-9]+[0-9]*|\./) && ($4 ~/[A-Za-z]+[1-9]+[0-9]*|\./)) {print NR}')
if [ ${#BAD_CHR_ROWS} -eq 0 ]; then
    echo "All rows have correct format for CHROMOSOME fields."
else
    echo "These rows have INCORRECT format for CHROMOSOME fields: ${BAD_CHR_ROWS%,*}."
fi

BAD_START_END_ROWS=$(zcat ${INPUT_BEDPE} | awk -F'\t' 'BEGIN { ORS="," } !(($2 ~/^([0-9]+)$|\-1/) && ($3 ~/^([0-9]+)$|\-1/) && ($5 ~/^([0-9]+)$|\-1/) && ($6 ~/^([0-9]+)$|\-1/)) {print NR}')
if [ ${#BAD_START_END_ROWS} -eq 0 ]; then
    echo "All rows have correct format for START and END fields."
else
    echo "These rows have INCORRECT format for START and END fields: ${BAD_START_END_ROWS%,*}."
fi

BAD_MAPQ_STRAND_ROWS=$(zcat ${INPUT_BEDPE} | awk -F'\t' 'BEGIN { ORS="," } !(($8 >= 0 && $8 < 2^8) && ($9 >= 0 && $9 < 2^8) && ($10 == "-" || $10 == "+" || $10 == ".") && ($11 == "-" || $11 == "+" || $11 == ".")) {print NR}')
if [ ${#BAD_MAPQ_STRAND_ROWS} -eq 0 ]; then
    echo "All rows have correct format for MAPQ and STRAND fields."
else
    echo "These rows have INCORRECT format for MAPQ and STRAND fields: ${BAD_MAPQ_STRAND_ROWS%,*}."
fi

BAD_CIGAR_ROWS=$(zcat ${INPUT_BEDPE} | awk -F'\t' 'BEGIN { ORS="," } !(($12 ~ /*|([0-9]+[MIDNSHPX=])+/) && ($13 ~ /*|([0-9]+[MIDNSHPX=])+/)) {print NR}')
if [ ${#BAD_CIGAR_ROWS} -eq 0 ]; then
    echo "All rows have correct format for CIGAR fields."
else
    echo "These rows have INCORRECT format for CIGAR fields: ${BAD_CIGAR_ROWS%,*}."
fi

BAD_FLAG_ROWS=$(zcat ${INPUT_BEDPE} | awk -F'\t' 'BEGIN { ORS="," } !(($14 >= 0 && $14 < 2^16) && ($15 >= 0 && $15 < 2^16)) {print NR}')
if [ ${#BAD_FLAG_ROWS} -eq 0 ]; then
    echo "All rows have correct format for FLAG field."
else
    echo "These rows have INCORRECT format for FLAG field: ${BAD_FLAG_ROWS%,*}."
fi

BAD_TLEN_ROWS=$(zcat ${INPUT_BEDPE} | awk -F'\t' 'BEGIN { ORS="," } !(($16 ~ /^(-*[0-9]+)$/) && ($17 ~ /^(-*[0-9]+)$/)) {print NR}')
if [ ${#BAD_TLEN_ROWS} -eq 0 ]; then
    echo "All rows have correct format for TLEN field."
else
    echo "These rows have INCORRECT format for TLEN field: ${BAD_TLEN_ROWS%,*}."
fi

BAD_NM_ROWS=$(zcat ${INPUT_BEDPE} | awk -F'\t' 'BEGIN { ORS="," } !(($18 ~ /NA|^([0-9]+)$/) && ($19 ~ /NA|^([0-9]+)$/)) {print NR}')
if [ ${#BAD_NM_ROWS} -eq 0 ]; then
    echo "All rows have correct format for NM field."
else
    echo "These rows have INCORRECT format for NM field: ${BAD_NM_ROWS%,*}."
fi


EMPTY_FIELDS=($(zcat ${INPUT_BEDPE} | awk -F'\t' 'BEGIN { ORS=" " } {for(N=1; N<=NF; N++) if($N=="" || $N ~/\s+/) print NR}'))
EMPTY_FIELDS_UNIQ_ROWS=$(printf "%s\n" "${EMPTY_FIELDS[@]}" | sort -V -u | tr '\n' ',')
EMPTY_FIELDS_UNIQ_ROWS2=${EMPTY_FIELDS_UNIQ_ROWS%,*}
if [ ${#EMPTY_FIELDS_UNIQ_ROWS2} -eq 0 ]; then
    echo "All rows have NO EMPTY or field with only WHITESPACES."
else
    echo "These rows have cells that are EMPTY or contain only WHITESPACES: ${EMPTY_FIELDS_UNIQ_ROWS%,*}."
fi


time2=$(date +%s)
echo "Job ended at "$(date) 
echo "Job took $(((time2-time1)/3600)) hours $((((time2-time1)%3600)/60)) minutes $(((time2-time1)%60)) seconds"

## EOF
